### ISSUE TITLE
Histogram of Oriented Gradients (KNN etc.)

### ISSUE THREAD

**Issue by jonasbarth** [jonasbarth/fds-2022-final-project#7]
# What we want to do?
Create histogram of oriented gradients and use as input to a classifier, starting with KNN.

# How

## Preprocessing
The goal of the preprocessing is to create a `Pipeline` object that we can reuse when using the test data. For every image:
1. Load image
2. Downsample (optional) (try 75%, 50%, 25% and evaluate)
3. Apply Gaussian blur (from opencv or scikit learn)
4. Create histogram of oriented gradients using `hog` from `skimage`  for a selection of channels. The output will be a single 2D matrix. (play around with the parameters, pixels_per_cell, cells_per_block) [documentation](https://scikit-image.org/docs/stable/api/skimage.feature.html#skimage.feature.hog). It IS possible to flatten the image using the boolean `feature_vector` parameter.
7. Save histograms in separate directory.

## Evaluation
1. Apply pipeline to test data
2. Get scores for test data, using confusion matrix from sklearn.
3. Select different features (channels) and retry. We can use cross validation (using sklearn).

## Visualisations
Somehow show the scores of the model. Maybe use the same plot as in the paper
![image](https://user-images.githubusercontent.com/32641360/206516872-621ac121-393e-4c87-a4fa-d02f9f0405e5.png)

**Comment by Mamiglia** [comment]
While I was browsing Sklearn documentation I came across the [KBinsDiscretizer](
https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.KBinsDiscretizer.html#sklearn.preprocessing.KBinsDiscretizer), which looks like exactly what we need to build histograms: 

Also this [article](https://towardsdatascience.com/step-by-step-tutorial-of-sci-kit-learn-pipeline-62402d5629b6) might be useful to understand pipelines @MattiaCastaldo

**Comment by jonasbarth** [comment]
First Results:

| channels | n neighbours | orientations | pixels per cell | cells per block | accuracy |
| -------- | ------------ | ------------ | --------------- | --------------- | -------- |
| 1,2,3    | 3            | 8            | (16, 16)        | (1, 1)          | 0.49     |
| 1,2,3    | 5            | 8            | (16, 16)        | (1, 1)          | 0.51     |
| 1,2,3    | 7            | 8            | (16, 16)        | (1, 1)          | 0.54     |
| 6,7,8    | 3            | 8            | (16, 16)        | (1, 1)          | 0.46     |
| 6,7,8    | 5            | 8            | (16, 16)        | (1, 1)          | 0.5      |
| 6,7,8    | 7            | 8            | (16, 16)        | (1, 1)          | 0.48     |
| 13,14,15 | 3            | 8            | (16, 16)        | (1, 1)          | 0.47     |
| 13,14,15 | 5            | 8            | (16, 16)        | (1, 1)          | 0.47     |
| 13,14,15 | 7            | 8            | (16, 16)        | (1, 1)          | 0.5      |

We need to play with the hog parameters:
* orientations
* pixels per cell
* cells per block

**Comment by jonasbarth** [comment]
Actually, it looks like the data I was working with was wrong and/or incomplete. We had a bug in the preprocessing which gave us a lot of empty images. That is now fixed and the scores are much better:

![hog_knn_confusion_matrix](https://user-images.githubusercontent.com/32641360/208059782-b0cb1678-5b09-426d-a8db-ec0f964c55b5.jpg)

**Comment by martinezvelascojavier** [comment]
When training a Logistic Regression model with HOGs, we find out that, whereas channels RGB and 6,7,8 fail to converge, channels 10,11 and 12 have an "almost perfect" performance: 
<img width="884" alt="image" src="https://user-images.githubusercontent.com/113454044/208139418-3b7ca923-9bf2-4e7c-955d-5dbd61e34769.png">
Looks like very good news :D

**Comment by Mamiglia** [comment]
This is so cool!
